orca/producer.go

```diff
@@ -0,0 +1,219 @@
+	// may be provided more frequently if another subscriber requests a shorter
+	// interval.
+	ReportInterval time.Duration
+}
+
+// RegisterOOBListener registers an out-of-band load report listener on sc.
+// Any OOBListener may only be registered once per subchannel at a time.  The
+// returned stop function must be called when no longer needed.
+func RegisterOOBListener(sc balancer.SubConn, l OOBListener, opts OOBListenerOptions) (stop func()) {
+	pr, close := sc.GetOrBuildProducer(producerBuilderSingleton)
+	p := pr.(*producer)
+	p.registerListener(l, opts.ReportInterval)
+
+	// TODO: When we can register for SubConn state updates, don't call run()
+	// until READY and automatically call stop() on SHUTDOWN.
+
+	// If stop is called multiple times, prevent it from having any effect on
+	// subsequent calls.
+	var once sync.Once
+	return func() {
+		once.Do(func() {
+			p.unregisterListener(l, opts.ReportInterval)
+			close()
+		})
+	}
+}
+
+type producer struct {
+	client v3orcaservicegrpc.OpenRcaServiceClient
+
+	ctx     context.Context
+	cancel  func()
+	closed  *grpcsync.Event // fired when closure completes
+	backoff func(int) time.Duration
+
+	mu        sync.Mutex
+	intervals map[time.Duration]int    // map from interval time to count of listeners requesting that time
+	listeners map[OOBListener]struct{} // set of registered listeners
+}
+
+// registerListener adds the listener and its requested report interval to the
+// producer.
+func (p *producer) registerListener(l OOBListener, interval time.Duration) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.listeners[l] = struct{}{}
+	p.intervals[interval]++
+}
+
+// registerListener removes the listener and its requested report interval to
+// the producer.
+func (p *producer) unregisterListener(l OOBListener, interval time.Duration) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	delete(p.listeners, l)
+	p.intervals[interval]--
+	if p.intervals[interval] == 0 {
+		delete(p.intervals, interval)
+	}
+}
+
+// minInterval returns the smallest key in p.intervals.
+func (p *producer) minInterval() time.Duration {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	min := time.Duration(-1)
+	for t := range p.intervals {
+		if min == time.Duration(-1) || t < min {
+			min = t
+		}
+	}
+	return min
+}
+
+// run manages the ORCA OOB stream on the subchannel.
+func (p *producer) run(ctx context.Context) {
+	defer p.closed.Fire()
+	backoffAttempt := 0
+	backoffTimer := time.NewTimer(0)
+	for ctx.Err() == nil {
+		select {
+		case <-backoffTimer.C:
+		case <-ctx.Done():
+			return
+		}
+
+		resetBackoff, err := p.runStream(ctx)
+
+		if resetBackoff {
+			backoffTimer.Reset(0)
+			backoffAttempt = 0
+		} else {
+			backoffTimer.Reset(p.backoff(backoffAttempt))
+			backoffAttempt++
+		}
+
+		switch {
+		case err == nil:
+			// No error was encountered; restart the stream.
+		case ctx.Err() != nil:
+			// Producer was stopped; exit immediately and without logging an
+			// error.
+			return
```

Comment:
How do we communicate this error to the user of this API? It will be useful for the LB policy to know about this error, right? So, that it can make other plans on how to balance load across the backends.

Reply:
This particular case isn't an error.  This indicates the user stopped the listener by calling `stop()`.

In the case where there is a stream error, we just log, which is part of the design.  Also included in the design but not here is logging a subchannel event in channelz.  We don't currently have any way to do that, so I'm omitting it for now.

Reply:
Maybe a TODO for the subchannel event in channelz?